ethcore/private-tx/src/state_store.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright 2015-2019 Parity Technologies (UK) Ltd.
+// This file is part of Parity Ethereum.
+
+// Parity Ethereum is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Ethereum is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Ethereum.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+use parking_lot::RwLock;
+use ethereum_types::H256;
+use kvdb::KeyValueDB;
+use types::transaction::SignedTransaction;
+use private_transactions::VerifiedPrivateTransaction;
+use private_state_db::PrivateStateDB;
+
+/// State of the private state sync
+#[derive(Clone, PartialEq)]
+pub enum SyncState {
+	/// No sync is running
+	Idle,
+	/// Private state sync is running
+	Syncing,
+}
+
+/// Wrapper over storage for the private states
+pub struct PrivateStateStorage {
+	verification_requests: RwLock<Vec<Arc<VerifiedPrivateTransaction>>>,
```

Comment:
I don't think you actually need so many `RwLock`s - all the data seems to be logically connected. Like in this code:
```Rust
(1)		self.state_storage.state_sync_completed(hash);
(2)		if self.state_storage.current_sync_state() == SyncState::Idle {
			trace!(target: "privatetx", "Private state sync completed, processing pending requests");
(3)			let creation_requests = self.state_storage.drain_creation_queue();
```
there are 3 calls which are using 3 different `RwLock`. And, for example, if check (2) passes (i.e. the state is `Idle`) and before line (3) someone adds more states to sync, `drain_*()` calls could actually return these new requests, for which states are missing => tx sign/verification will fail.

So probably reconsider API to avoid such issues (and leave <= 1 internal lock)?

Reply:
Reworked. 2 locks are left. The second one is needed for stalling requests logic implementation (in order to keep currently processed hashes in the separate structure)